Token Map: Introduce an efficient lookup and translation class for string mappings.

dmsnell opened the pull request:
Trac ticket: Core-60698
(previously Core-60841)

Motivated by the need to properly transform HTML named character references (like &amp;) I found the need for a new semantic class which can efficiently perform search and replacement of a set of static tokens. Existing patterns in the codebase are not sufficient for the HTML need, and I suspect there are other use-cases where this class would help.

---

In #6387 I have built a spec-compliant HTML5 text decoder utilizing this token map. The performance of the new decoder is approximately 20% slower than calling `html_entity_decode()` directly, except that it properly decodes what PHP can't. In fact, the performance bottleneck in that PR comes from converting into UTF-8 the sequence of digits in numeric character references, not in looking up named character references.

---

This proposal is adding a new class `WP_Token_Map` providing at least two methods for normal use:

 - `contains( $token )` returns whether the passed string is in the set.
 - `read_token( $text, $offset = 0, $skip_bytes )` indicates if the character sequence starting at the given offset in the passed string forms a token in the set, and if so, returns the replacement for that token. It also sets &$skip_bytes to the length of the token so that calling code .

It also provides utility functions for pre-computing these classes, as they are designed for relatively-static cases where the actual code is intended to be generated dynamically, but stay static over time. For example, HTML5 defines the set of named character references and indicates that the list shall not change or be expanded. [HTML5 spec](https://html.spec.whatwg.org/#named-character-references-table). Precomputing can save on the startup-up cost of building the optimized lookup tables.

 - `WP_Token_Map::from_array( array $mappings )` generates a new token map from the given array of whose keys are tokens and whose values are the replacements.
 - `to_array()` dumps the set of mapping into an array suitable for passing back into `from_array()`.
 - `WP_Token_Map::from_precomputed_table( ...$table )` instantiates a token set from a precomputed table, skipping the computation for building the table and sorting the tokens.
 - `precomputed_php_source_table()` generates PHP source code which can be loaded with the previous static method for maintenance of the core static token sets.

#### Other potential uses
 - Converting smilies like :)
 - Converting emoji sequences like :happy:
 - Determining if a given verb/action is allowed in an API call.

dmsnell commented:
Rebased from e98508f10d17b606cf3f8761c82bb0f8750b263d to 502e4f5bcbe1271d57efe81aebe4ac19884e6cb7

dmsnell committed:
Introduce Token Map: An optimized static translation class.

This patch introduces a new class: `WP_Token_Map`, which is designed for
efficient lookup and translation of static mappings between string keys
or tokens, and string replacements (for example, HTML character
references).

The token map incorporates certain limitations on the string length of
the tokens, but takes advantage of multiple optimizations to efficiently

dmsnell committed:
Add case-insensitivity and early-abort for small words when none exist.

gziolo commented on tests/phpunit/tests/wp-token-map/wpTokenMap.php:
```diff
@@ -0,0 +1,377 @@
+
+			case 'HTML5':
+				if ( ! isset( $html5_character_references ) ) {
+					$dataset = json_decode( file_get_contents( __DIR__ . '/../../data/html5-entities.json' ), JSON_OBJECT_AS_ARRAY ); // phpcs:ignore.
```
There is `wp_json_file_decode` helper method that could be used.

dmsnell replied:
used it! makes the testing a bit more annoying locally because I had to polyfill that function in my `bootstrap.php`, but I like the built-in error-checking in the helper.

gziolo commented on tests/phpunit/tests/wp-token-map/wpTokenMap.php:
```diff
@@ -0,0 +1,377 @@
+		return $html5_token_map;
+	}
+
+	const ANIMAL_EMOJI = array(
```
Here it is. I was searching for it at the top of the file 😄

dmsnell replied:
can be found at the top now!

gziolo commented on tests/phpunit/tests/html-api/generate-html5-named-character-references.php:
```diff
@@ -0,0 +1,78 @@
+ *         'characters' => '©',
+ *     );
+ *
+ * @see https://html.spec.whatwg.org/entities.json
```
It would be helpful to explicitly explain in the code comment that the list is copied from the provided URL. It took me some time to understand how everything is assembled. Should the script also try to update the list of HTML5 entities by fetching the most recent data from the HTML spec?

sirreal replied:
I was going to suggest something similar. Maybe the entities data file should go in a directory with a README that can links to https://html.spec.whatwg.org/entities.json.

dmsnell replied:
I've added a comment. Can leave it open to where it belongs, as I had almost zero sense of where that should be.

I don't like it in the `src` directory because it isn't source code in the same sense. Some test directory?

sirreal replied:
I think the data file is in the right place. My recommendation was something like this:

- tests/phpunit/data/html5-entities/entities.json <- the file
- tests/phpunit/data/html5-entities/README.md <- explanation

This is similar to https://github.com/WordPress/wordpress-develop/tree/19faad774454e86acdd1f0549816cff22b0eb9d8/tests/phpunit/data/html5lib-tests that has a readme with relevant information about the data included.

gziolo replied:
Sounds like a good plan. You can also reference the README in the script that generates the file.

dmsnell replied:
README added and directory restructured.

gziolo commented on tests/phpunit/tests/wp-token-map/wpTokenMap.php:
```diff
@@ -0,0 +1,377 @@
+	 *
+	 * @return WP_Token_Map
+	 */
+	public static function get_html5_token_map() {
```
Is there any convention when these static helpers should be private vs public? `get_test_input_array` is private, and this helper is public.

dmsnell replied:
They should all be private; they are all now private. Thanks!

gziolo commented on src/wp-includes/class-wp-token-map.php:
```diff
@@ -0,0 +1,728 @@
+ *      2    === $bytes_skipped;
+ *
+ *      echo $smilies->precomputed_php_source_table( '    ' );
+ *      // Output.
```
Nit: it took me some time to understand what `echo $smilies->precomputed_php_source_table( '    ' );` does. Maybe the comment could get updated here and in other places to:
```suggestion
 *      // Output:
```

dmsnell replied:
I've expanded the comment and rearranged the example. Mind giving it another lookover?

gziolo replied:
That is crystal clear now 👍

gziolo commented on src/wp-includes/class-wp-token-map.php:
```diff
@@ -0,0 +1,728 @@
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Token Map tokens and substitutions must all be shorter than 256 bytes.' ),
+					'6. .0'
```
Looks like typo:
```suggestion
					'6.6.0'
```

dmsnell replied:
Fixed, thanks!

gziolo commented on tests/phpunit/tests/wp-token-map/wpTokenMap.php:
```diff
@@ -0,0 +1,377 @@
+	 * @return array[].
+	 */
+	public static function data_html5_test_dataset() {
+		$html5 = self::get_test_input_array( 'HTML5' );
```
Would it be valuable to add a test that validates that `$html5` contains a certain number of tokens? Maybe it's impossible, but if for some reason the tokens wouldn't be loaded from the data file, then non tests depending on the data provider would run.

dmsnell replied:
I did look into adding `count()` just for the sake of the tests, but then wasn't sure it was necessary for the runtime and it was slightly complicated.

in the case of failure to load, I think we would have a backup, which is a disagreement between the html5 class and the entities JSON file.

also, surprisingly I was able to turn the last data provider non-static and add an assertion, so I've made a constant and verified the count that way, as a way to ensure the validity of the test without making the tests lead the design of the class.

gziolo replied:
Running the assertion while reading the test data fits perfectly 💯

I wasn't sure about such usage in the data provider, but it's just a helper so
seems to work great.

gziolo reviewed (commented):
I did some high level pass on the PR and left my feedback. I was looking at the code from the HTML API encoding/decoding angle. I don't have better insights to share than @aaronjorbin in the Trac ticket.

Overall, the code is very complex and very specific to the requirements that HTML API has so I'm not the best person to draw any conclusions on whether this class is general enough. However, based on the explanation provided by @dmsnell, it definitely is important part of the implementation in the terms of performance for the work on HTML entities encoding and decoding.

dmsnell committed:
PR Feedback

dmsnell committed:
Verify count of named character references.

gziolo commented:
Some CI jobs report that some static test helpers are private and it can’t access them.

sirreal commented on tests/phpunit/tests/html-api/generate-html5-named-character-references.php:
```diff
@@ -0,0 +1,82 @@
+<?php
+
+/**
+ * Auto-generated class for looking up HTML named character references.
+ *
+ * To regenerate, run the generation script directly.
+ *
+ * Example:
+ *
+ *     php tests/phpunit/tests/html-api/generate-html5-named-character-references.php
+ *
+ * @package WordPress
+ * @since 6.6.0
```
This is clear… as long as you stop to pay attention and read it. I'd consider doing something _louder_:

```suggestion
 * Auto-generated class for looking up HTML named character references.
 *
 * ⚠️ !!! THIS ENTIRE FILE IS AUTOMATICALLY GENERATED !!! ⚠️
 * Do not modify this file directly.
 *
 * To regenerate, run the generation script directly.
 *
 * Example:
 *
 *     php tests/phpunit/tests/html-api/generate-html5-named-character-references.php
 *
 * @package WordPress
 * @since 6.6.0
```

dmsnell replied:
Shouting now in d08d50f5b1

sirreal commented on tests/phpunit/tests/html-api/generate-html5-named-character-references.php:
```diff
@@ -0,0 +1,82 @@
+
+// phpcs:disable
+
+global \$html5_named_character_references;
```
Note for other reviewers:

It took me a while to see that `\$var` is escaping the `$` in `$var`. I kept seeing this as the namespace global space. In the generated file is just `global $html5_named_character_references;`

dmsnell replied:
Added a note in a comment above the template. Hope that helps

sirreal commented on tests/phpunit/tests/wp-token-map/wpTokenMap.php:
```diff
@@ -0,0 +1,404 @@
+		$html5 = self::get_test_input_array( 'HTML5' );
+		$cases = array();
+
+		$this->assertSame(
+			self::KNOWN_COUNT_OF_ALL_HTML5_NAMED_CHARACTER_REFERENCES,
+			count( $html5 ),
+			'Found the wrong number of HTML5 named character references: confirm the entities.json file."'
+		);
```
It seems out of place to have an assertion in the data provider. Maybe another simple test could not annotate the data provider, but call it make an assertion about its length?

dmsnell replied:
🤷‍♂️ maybe, thought then the test isn't testing the unit under test but the test runner itself. it's weird in the data provider, but does it warrant its own separate test?

I thought about `throw`ing in here too but then figured an assertion could be clearer.

I don't have strong feelings about this; the assertion inside the data provider seemed a fitting tool but I'm happy to reconsider.

sirreal commented on tests/phpunit/tests/wp-token-map/wpTokenMap.php:
```diff
@@ -0,0 +1,404 @@
+			'Found the wrong number of HTML5 named character references: confirm the entities.json file."'
+		);
+
+		foreach ( $html5 as $token => $replacement ) {
+			$cases[ $token ] = array( $token, $replacement );
+		}
+
+		return $cases;
```
This seems like an ideal place to yield cases, we don't need to collect and return them all at the same time:

```suggestion
		foreach ( $html5 as $token => $replacement ) {
			yield $token => array( $token, $replacement );
		}
```

The `$cases = array();` line above could also be removed.

dmsnell replied:
Great suggestion. Yielding in 2cf613f866c1321983f5f4385c0334585a43d75c

dmsnell committed:
Shallow abstractions diverging from their truth, tsk tsk

sirreal commented on tests/phpunit/tests/wp-token-map/wpTokenMap.php:
```diff
@@ -0,0 +1,404 @@
+
+			case 'HTML5':
+				if ( ! isset( $html5_character_references ) ) {
+					$dataset = wp_json_file_decode( __DIR__ . '/../../data/html5-entities.json', true );
```
[According to the docs this signature is wrong?
](https://developer.wordpress.org/reference/functions/wp_json_file_decode/)

Was this intended?

```suggestion
					$dataset = wp_json_file_decode( __DIR__ . '/../../data/html5-entities.json', array( 'associative' => true ) );
```

dmsnell replied:
yes, and I fixed it. shallow abstractions making their own language…

sirreal replied:
Isn't it still wrong?

https://github.com/WordPress/wordpress-develop/blob/d63da61482e77d4c031f80fca201248b5080b4f7/tests/phpunit/tests/wp-token-map/wpTokenMap.php#L63

The second argument should be an array with a single key: `associative`. It's the value `true`.

dmsnell committed:
Restructure generator script

dmsnell committed:
fixup! Restructure generator script

sirreal commented on tests/phpunit/data/html5-entities/generate-html5-named-character-references.php:
```diff
@@ -0,0 +1,88 @@
+<?php
```
Nice, this is a good place for this script too 👍

dmsnell committed:
PR Feedback

sirreal commented on tests/phpunit/data/html5-entities/README.md:
````diff
@@ -0,0 +1,22 @@
+
+## Updating the optimized lookup class.
+
+The `html5-named-character-references.php` file contains an optimized lookup map for the entities in `entities.json`.
+Run the `generate-html5-named-character-references.php` file to update the auto-generated Core module.
+
+```bash
+~$ php tests/phpunit/data/html5-entities/generate-html5-named-character-references.php
+OK: Successfully generated optimized lookup class.
+```
````
This is super helpful.

Maybe we could make it a bit clearer exactly what file this is pointing to and even link it (I _think_ that's correct syntax):

````suggestion
The [`html5-named-character-references.php`][1] file contains an optimized lookup map for the entities in `entities.json`.
Run the `generate-html5-named-character-references.php` file to update the auto-generated Core module.

```bash
~$ php tests/phpunit/data/html5-entities/generate-html5-named-character-references.php
OK: Successfully generated optimized lookup class.
```

- [1]: ../../../../src/wp-includes/html-api/html5-named-character-references.php
````

dmsnell committed:
Update tests/phpunit/data/html5-entities/README.md

Co-authored-by: Jon Surrell <sirreal@users.noreply.github.com>

dmsnell committed:
Fix markdown link reference

dmsnell committed:
Add more links

sirreal commented on src/wp-includes/class-wp-token-map.php:
```diff
@@ -0,0 +1,740 @@
+	private $key_length = 2;
+
+	/**
+	 * Stores an optimized form of the word set, where words are grouped
```
I'd like to see the explicit rule for large vs. small words in their docs:

> …where large words are strings whose strlen greater than is key_length

(that's almost certainly wrong, just an example)

dmsnell replied:
added class-level description in [31c57b0](https://github.com/WordPress/wordpress-develop/pull/5373/commits/31c57b0871a193fdf18a6ae870cb32b0750799a2)

sirreal commented on src/wp-includes/class-wp-token-map.php:
```diff
@@ -0,0 +1,740 @@
+	private $groups = '';
+
+	/**
+	 * Stores an optimized row of small words, where every entry is
```
I'd like to see the explicit rule for large vs. small words in their docs:

> …where small words are strings whose strlen is less than or equal to key_length

(that's almost certainly wrong, just an example)

dmsnell replied:
added a class-level description in [31c57b0](https://github.com/WordPress/wordpress-develop/pull/5373/commits/31c57b0871a193fdf18a6ae870cb32b0750799a2)

sirreal reviewed (commented):
This is great, but there's a lot in here. I'm digging into the implementation more now and I'll finish a complete review tomorrow.

dmsnell committed:
I'm no longer a fan of this shallow abstraction

dmsnell committed:
Define large and small words in class doc.

dmsnell committed:
Make test data provider public where necessary

sirreal committed:
Remove "." from PHPUnit ticket annotations

The ticket annotation also creates a group for filtering PHPUnit tests.
When the annotation is terminated in ".", filtering tests like
"--group=60698" does not work as expected.

sirreal committed:
Fix casing of class in coversDefaultClass annotation

sirreal commented on tests/phpunit/tests/wp-token-map/wpTokenMap.php:
```diff
@@ -0,0 +1,404 @@
+			$datasets[ $dataset_name ] = array( self::get_test_input_array( $dataset_name ) );
+		}
+
+		return $datasets;
```
This is another place where yielding is appropriate:

```suggestion
		foreach ( $dataset_names as $dataset_name ) {
			yield $dataset_name => array( self::get_test_input_array( $dataset_name ) );
		}
```

sirreal commented on tests/phpunit/tests/wp-token-map/wpTokenMap.php:
```diff
@@ -0,0 +1,404 @@
+ * @package WordPress
+ *
+ * @since 6.6.0
+ *
```
Should this have a group annotation?

A ticket annotation can be added here as well.

dmsnell replied:
I wasn't sure of any group to use. When I added this I poked around at some other top-level modules in `wp-includes` and didn't see an obvious group. Happy to take suggestions.

aaronjorbin replied:
I think we can add a new `token-map` group.  It also might be worth putting in the `html-api` group since this is where the bigger effort lies

sirreal commented on src/wp-includes/class-wp-token-map.php:
```diff
@@ -0,0 +1,751 @@
+	 *
+	 * @return WP_Token_Map Map with precomputed data loaded.
+	 */
+	public static function from_precomputed_table( $key_length, $groups, $large_words, $small_words, $small_mappings ) {
```
Have you thought about versioning here? What would happen if we change internal data or the way that precomputed data is generated?

dmsnell replied:
I've thought about it, and now that you bring it up, I think it would be best to eagerly include a version so that we can ignore potential conflicts. I was trying to avoid creating a separate state value too, but it may help with versioning to only have a single exported var as an array. I'll rework this.

sirreal commented on src/wp-includes/class-wp-token-map.php:
```diff
@@ -0,0 +1,751 @@
+	 *
+	 * @since 6.6.0
+	 *
+	 * @param string  $text              String in which to search for a lookup key.
+	 * @param ?int    $offset            How many bytes into the string where the lookup key ought to start.
+	 * @param ?int    &$skip_bytes       Holds byte-length of found lookup key if matched, otherwise not set.
+	 * @param ?string $case_sensitivity 'case-insensitive' to ignore ASCII case or default of 'case-sensitive'.
+	 * @return string|false Mapped value of lookup key if found, otherwise `false`.
+	 */
+	public function read_token( $text, $offset = 0, &$skip_bytes = null, $case_sensitivity = 'case-sensitive' ) {
```
This documentation makes it seem like `$skip_bytes` is optional, but I tried to use this to write a test while setting case insensitive and could not:

```php
// $map is valid
$map->read_token( 'x', 0, null, 'case-insensitive' );
// Error: WP_Token_Map::read_token(): Argument #3 ($skip_bytes) could not be passed by reference
```

Is the function intended to be called like that? Is there a way to set case insensitive if I don't care about the `$skip_bytes`?

dmsnell replied:
Note privately: I kind of added this as an experiment while testing, but like it. It's mainly ASCII-oriented, like most basic string casing functions, which seems fine given its role. For example, finding upper or lowercase tag names. I'll clarify in the documentation.

dmsnell replied:
Yes it's supposed to be optional and is, but I think you're running into issues with PHP and trying to pass an argument after an optional one. I think if you want to pass a final argument you will need to supply a real variable for `&$skip_bytes`. Maybe this is a quirk to handle, but I don't think it's particularly an outpouring of the design of this function as much as it derives from PHP.

sirreal replied:
([this is a reply to other comments](https://github.com/WordPress/wordpress-develop/pull/5373#discussion_r1605331487) - GitHub doesn't really thread well when it's part of a review)

For the record, it seems you can pass an uninitialized variable name to the function and it's not even a warning in this case:

```php
// first place $skip_bytes appears
$map->read_token( 'x', 0, $skip_bytes, 'case-insensitive' );
var_dump($skip_bytes); // int(0)
```

dmsnell replied:
this is another funny way PHP works. It's only code linters that require a previous declaration of a variable in PHP, not the language. it's nice seeing a variable declared ahead of time, but completely pointless apart from communicating to other devs.

sirreal reviewed (commented):
I've been reviewing the implementation and I have a concern. As far as I can tell, case-insensitive behavior is unreliable on some string strings, specifically multibyte strings with upper/lowercase variants like `ü`.

<details>
<summary>diff with broken case-insensitive match tests</summary>

```diff
diff --git before/tests/phpunit/tests/wp-token-map/wpTokenMap.php after/tests/phpunit/tests/wp-token-map/wpTokenMap.php
index 8b28982a64..56331f2e93 100644
--- before/tests/phpunit/tests/wp-token-map/wpTokenMap.php
+++ after/tests/phpunit/tests/wp-token-map/wpTokenMap.php
@@ -401,4 +401,48 @@ class Tests_WpTokenMap extends WP_UnitTestCase {
 
 		return $html5_token_map;
 	}
+
+	/**
+	 */
+	public function test_mb_matches_short() {
+		$small_u_umlaut = 'ü';
+		$big_u_umlaut   = 'Ü';
+		$map            = WP_Token_Map::from_array(
+			array(
+				$big_u_umlaut => 'upper U-umlaut',
+			)
+		);
+
+		// Ensure the strings are "small" for the token map
+		$this->assertLessThanOrEqual( 2, strlen( $small_u_umlaut ) );
+		$this->assertLessThanOrEqual( 2, strlen( $big_u_umlaut ) );
+
+		$this->assertNotSame( $small_u_umlaut, $big_u_umlaut );
+		$this->assertSame( mb_strtoupper( $small_u_umlaut ), $big_u_umlaut );
+
+		$this->assertTrue( $map->contains( $small_u_umlaut, 'case-insensitive' ) );
+		$skip_bytes = 0;
+		$this->assertSame( 'upper U-umlaut', $map->read_token( $small_u_umlaut, 0, $skip_bytes, 'case-insensitive' ) );
+	}
+
+
+	/**
+	 */
+	public function test_mb_matches_long() {
+		$small_x_with_dot = 'ẋ';
+		$big_x_with_dot   = 'Ẋ';
+		$map              = WP_Token_Map::from_array(
+			array( $big_x_with_dot => 'upper x with dot above' ),
+		);
+
+		// Ensure the strings are "big" for the token map
+		$this->assertGreaterThan( 2, strlen( $small_x_with_dot ) );
+		$this->assertGreaterThan( 2, strlen( $big_x_with_dot ) );
+
+		$this->assertNotSame( $small_x_with_dot, $big_x_with_dot );
+		$this->assertSame( mb_strtoupper( $small_x_with_dot ), $big_x_with_dot );
+
+		$this->assertTrue( $map->contains( $small_x_with_dot, 'case-insensitive' ) );
+		$skip_bytes = 0;
+		$this->assertSame( 'upper x with dot above', $map->read_token( $small_x_with_dot, 0, $skip_bytes, 'case-insensitive' ) );
+	}
 }
```

</details>

Have you considered this? 

Some options that come to mind:
- Document that it may not work under some conditions.
- Set a flag on the token map and perform checks in `from_array` to set the flag if multibyte chars with upper/lower variants are detected. Store and load the flag from precomputed. Warn in `from_array`? Throw if case-insensitive is used with the flag set?
- Fix it somehow? Maybe use that precomputed flag to know whether its necessary to use `mb_strtoupper` or `mb_strtolower` functions on the large and small words when doing case-insensitive comparison.
- Remove case-insensitive matching.

sirreal committed:
Use more verbose length variables

sirreal commented on src/wp-includes/class-wp-token-map.php:
```diff
@@ -0,0 +1,751 @@
+	 *
+	 * @param string $a First string to compare.
+	 * @param string $b Second string to compare.
+	 * @return int -1 if `$a` is less than `$b`; 1 if `$a` is greater than `$b`, and 0 if they are equal.
```
This returns `<0` or `>0`, but not strictly `-1` or `1`.

dmsnell replied:
Fixed in 42c0d085c000259752d9287172608d951b7a954a - I had copied the docs from `strcmp` and had no accounted for the return from strings of different lengths.

dmsnell committed:
Update tests/phpunit/tests/wp-token-map/wpTokenMap.php

Co-authored-by: Jon Surrell <sirreal@users.noreply.github.com>

dmsnell commented:
> I've been reviewing the implementation and I have a concern. As far as I can tell, case-insensitive behavior is unreliable on some string strings, specifically multibyte strings with upper/lowercase variants like ü.

> Have you considered this?

Yes, absolutely.

I did add case insensitivity during the testing cycle kind of as an afterthought, but I like having it. It's intended to aid in the purpose for which the class was made, which is focusing on ASCII casing. This may sound strange, but almost every basic casing functions are primarily if not entirely ASCII-oriented, because proper case-folding is a much more complicated topic, and no, `mb_strtoupper()` is not enough, at all.

So I think the expectation that this works for `ü` is more subjective. My gut feeling is that we can clarify this in documentation and let it be. There's no way this is going to be properly detecting more complicated case variants, but then again, nothing in this domain does. Otherwise we can remove it, but I like how the option gives us the ability to implement functions like `WP_HTML_Processor::is_special()` without allocating a new string and upper-casing it.

sirreal commented:
It seems fine to me if we don't support case insensitive non-ascii token matching as long as we're clear about it.

dmsnell committed:
Wrap precomputed state in a single array, add version.

dmsnell committed:
Fix translation placeholders

dmsnell commented:
> It seems fine to me if we don't support case insensitive non-ascii token matching as long as we're clear about it.

Ha! I went to add these and I already noted this in the function docblock.

> 'case-insensitive' to ignore ASCII case or default of 'case-sensitive'

<img width="563" alt="Screenshot 2024-05-17 at 11 28 38 AM" src="https://github.com/WordPress/wordpress-develop/assets/5431237/f9d3bf66-9307-4f67-b4bb-6e8f1aecacfa">

Text issues are apparently always on my mind. Let me know if you still think this is unclear. I'm worried that if we try and add more documentation when it's already there right at the cursor when calling the function, then it won't be any more helpful, just more noise.

dmsnell committed:
Fix string interpolation issue.

dmsnell committed:
Easier to be worse

dmsnell committed:
Update the HTML5 reference table.

sirreal commented:
> I already noted this in the function docblock … Let me know if you still think this is unclear.

I think this is fine. I was searching for words like "mb" or "multibyte", but it seems like ASCII is correct 👍 

I did notice a [changelog entry on the PHP documentation page](https://www.php.net/manual/en/function.strtoupper.php#refsect1-function.strtoupper-changelog) that makes me wonder if we might need to support PHP versions that have varying behavior on non-ASCII treatment. I wasn't able to trigger any differing behavior myself, but do you know what this is about?

> `8.2.0` Case conversion no longer depends on the locale set with [`setlocale()`](https://www.php.net/manual/en/function.setlocale.php). Only ASCII characters will be converted.

That seems like exactly what we want, but I've looked around PHP source, the changelog, etc. and I haven't managed to find exactly what changed or examples of the different.

sirreal commented on src/wp-includes/class-wp-token-map.php:
```diff
@@ -0,0 +1,810 @@
+		$ignore_case  = 'case-insensitive' === $case_sensitivity;
+		$small_length = strlen( $this->small_words );
+		$search_text  = substr( $text, $offset, $this->key_length );
+		if ( $ignore_case ) {
+			$search_text = strtoupper( $search_text );
+		}
```
I was comparing the performance of `strtoupper` and `strcasecmp` here. I thought that using `strcasecmp` would outperform `strtoupper`, but the results weren't conclusive. I suspect it's likely highly dependent on inputs.

**EDIT:** To be perfectly clear this is _not_ a change request. I'm documenting that I explored alternatives in the implementation.

<details>

<summary>patch</summary>

```diff
diff --git before/src/wp-includes/class-wp-token-map.php after/src/wp-includes/class-wp-token-map.php
index 40c0520659..75beabe074 100644
--- before/src/wp-includes/class-wp-token-map.php
+++ after/src/wp-includes/class-wp-token-map.php
@@ -575,19 +575,16 @@ class WP_Token_Map {
 	 * @return string|false Mapped value of lookup key if found, otherwise `false`.
 	 */
 	private function read_small_token( $text, $offset, &$skip_bytes, $case_sensitivity = 'case-sensitive' ) {
-		$ignore_case  = 'case-insensitive' === $case_sensitivity;
-		$small_length = strlen( $this->small_words );
-		$search_text  = substr( $text, $offset, $this->key_length );
-		if ( $ignore_case ) {
-			$search_text = strtoupper( $search_text );
-		}
+		$ignore_case   = 'case-insensitive' === $case_sensitivity;
+		$small_length  = strlen( $this->small_words );
+		$search_text   = substr( $text, $offset, $this->key_length );
 		$starting_char = $search_text[0];
 
 		$at = 0;
 		while ( $at < $small_length ) {
 			if (
 				$starting_char !== $this->small_words[ $at ] &&
-				( ! $ignore_case || strtoupper( $this->small_words[ $at ] ) !== $starting_char )
+				( ! $ignore_case || 0 !== strcasecmp( $this->small_words[ $at ], $starting_char ) )
 			) {
 				$at += $this->key_length + 1;
 				continue;
@@ -601,7 +598,7 @@ class WP_Token_Map {
 
 				if (
 					$search_text[ $adjust ] !== $this->small_words[ $at + $adjust ] &&
-					( ! $ignore_case || strtoupper( $this->small_words[ $at + $adjust ] !== $search_text[ $adjust ] ) )
+					( ! $ignore_case || 0 !== strcasecmp( $this->small_words[ $at + $adjust ], $search_text[ $adjust ] ) )
 				) {
 					$at += $this->key_length + 1;
 					continue 2;
```

</details>

dmsnell replied:
thanks for the alternative exploration. I am not sure it would be possible to use `strcasecmp` like this here because we aren't looking for full matches, necessarily, in the `small_words` array. we're really looking for a match in the `$text` of a string in the small words array, but the words in the small words array are zero-extended.

for example, if we have `ab` in `$text` at the `$offset`, and `"a\x00\x00"` in the small words array, we will not match `0 === strcasecmp()`, but there is indeed a match.

so we could add a length to the comparison, but then if it didn't match, we'd need to backtrack and try matching with `length - 1` and so on until `length === 1` and there are no matches.

by crawling forward only in the document we can ensure that any prefix of the input text will match as expected on a word shorter than the longest small word.

sirreal replied:
The idea is still to do character by character comparison, not full string comparison, but instead of calling `strtoupper` on the needle and on the characters, `strcasecmp` does the work of ASCII case-insensitive matching.

sirreal commented on src/wp-includes/class-wp-token-map.php:
```diff
@@ -0,0 +1,810 @@
+	 *
+	 * @param string  $text              String in which to search for a lookup key.
+	 * @param ?int    $offset            How many bytes into the string where the lookup key ought to start.
+	 * @param ?int    &$skip_bytes       Holds byte-length of found lookup key if matched, otherwise not set.
```
This parameter is very interesting. It seems very important, it's a bit like `$matches` in `preg_match` in that it's an output of the function and not an input.

I think the name `skip_bytes` can be improved. That's very focused on what calling code would likelye do with the value and less what the value _is_. Maybe something like `$matched_token_bytelength` or would be a descriptive name of what it is?

gziolo replied:
Yes, please. It was nearly impossible for me to understand the exact purpose of this `$skip_bytes` param by looking at code examples. I'm still not sure why it would be useful for developers.

sirreal replied:
The idea is that you're working with an input like this and using a token map to detect (and perform) replacements:

```html
true &not; false
```

We'd like to decode this by replacing entities, so we start by checking matches (this would be in a loop):

```php
$m = WP_Token_Map::from_array([ '&not;' => '¬' ]);
$s = 'true &not; false';

//  check here
//  |
//  v
// "true &not; false"
$replacement = $m->read_token( $s, 0, $skip_bytes );
// $replacement = false
// $skip_bytes = null

//   check here
//   |
//   v
// "true &not; false"
$replacement = $m->read_token( $s, 1, $skip_bytes );
// $replacement = false
// $skip_bytes = null

// …

//       check here
//       |
//       v
// "true &not; false"
$replacement = $m->read_token( $s, 5, $skip_bytes );
// $replacement = "¬"
// $skip_bytes = 5
```

At this point, we've hit a match and would perform replacement, but the function just returned `"¬"` which means "at index 5 there's a token matching `¬`" but we don't know what the token is so it's unclear how we should perform replacement. We need to know where it starts (we know the offset, we passed it in as an argument) what it matches (`¬` was returned) but also where the token ends. That's where `$skip_bytes` is involved, it lets us know the byte length of the matched token so we can replace (or resume looking after the end of the token).

```php
// 5 here is the index in the string where we found the token, we know that from our
// original iteration
substr( $s, 0, 5 ) . $replacement . substr( $s, 5 + $skip_bytes ); // "true ¬ false"
```

gziolo replied:
Ok, so it’s the length of the token matched that we will ignore while constructing the replaced version. The coincidence that the offset and skip bytes were both equal to 5 made this puzzle more challenging, but I hope I got it right. I was mostly making the case that it was the only difficult part to reason in the PHPDoc.

dmsnell replied:
sounds like I should pull in some of the wording from the dev note and move it into the docblocks…

https://make.wordpress.org/core/?p=113042&preview=1&_ppp=452181011b

sirreal replied:
> it’s the length of the token matched

Yep! In this case it's `strlen( "&not;" )`: 5.

> the coincidence that the offset and skip bytes were both equal to 5 made this puzzle more challenging.

🤦 yes, poor execution in my example 😓

dmsnell replied:
I've renamed the argument to `$matched_token_byte_length`. It's verbose, but more explanatory.

sirreal commented on src/wp-includes/class-wp-token-map.php:
```diff
@@ -0,0 +1,810 @@
+	 *
+	 * Example:
+	 *
+	 *     false === $smilies->read_token( 'Not sure :?.', 0, $bytes_skipped );
+	 *     '😕'  === $smilies->read_token( 'Not sure :?.', 9, $bytes_skipped );
+	 *     2     === $bytes_skipped;
+	 *
+	 * Example:
+	 *
+	 *     while ( $at < strlen( $input ) ) {
+	 *         $next_at = strpos( $input, ':', $at );
+	 *         if ( false === $next_at ) {
+	 *             break;
+	 *         }
+	 *
+	 *         $smily = $smilies->read_token( $input, $next_at, $bytes_skipped );
+	 *         if ( false === $next_at ) {
+	 *             ++$at;
+	 *             continue;
+	 *         }
+	 *
+	 *         $prefix  = substr( $input, $at, $next_at - $at );
+	 *         $at     += $bytes_skipped;
+	 *         $output .= "{$prefix}{$smily}";
+	 *     }
```
Note that this example uses `bytes_skipped` instead of `skip_bytes`, which is slightly confusing. It would be good to use the parameter name.

dmsnell replied:
they were different because of perspective. including `matched_` in the name I think helps the wording on the internal side, but I think that calling code still has a valid reason to use a different name. maybe I can further clarify with something like `$handle_length`

sirreal reviewed (approved):
I've been over this in detail and left a lot of feedback. Thanks for the patience @dmsnell. I've left what I believe are my final questions, comments, and suggestions, but I feel good about this change in general.

aaronjorbin commented on src/wp-includes/class-wp-token-map.php:
```diff
@@ -0,0 +1,810 @@
+			) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Token Map tokens and substitutions must all be shorter than 256 bytes.' ),
```
Minor: Rather than hardcoding 256, I think this should use `self::MAX_LENGTH` on the chance it changes in the future. 
```suggestion
					/* translators: maximum byte length */ 
					sprintf( __( 'Token Map tokens and substitutions must all be shorter than %i bytes.' ), self::MAX_LENGTH ),
```

dmsnell replied:
good catch! incorporated in 0e7ab4ca52d97da25a20e0e493bc33308f56cf1d

dmsnell committed:
Clarify case insensitivity.

dmsnell commented:
> I did notice a [changelog entry on the PHP documentation page](https://www.php.net/manual/en/function.strtoupper.php#refsect1-function.strtoupper-changelog) that makes me wonder if we might need to support PHP versions that have varying behavior on non-ASCII treatment. I wasn't able to trigger any differing behavior myself, but do you know what this is about?

I've struggled to test this with the HTML API. in effect, some system locales could cause, I believe, a few characters to be case-folded differently, but this is such a broken system I'm not of the opinion that it's a big risk here. I'm willing to let this sit as a hypothetical bug until we have a clearer understanding of it.

dmsnell committed:
Rename $skip_bytes to $matched_token_byte_length

dmsnell committed:
Fix compare result description.

dmsnell committed:
Update docblock wording and examples

aaronjorbin commented on tests/phpunit/tests/wp-token-map/wpTokenMap.php:
```diff
@@ -0,0 +1,400 @@
+	 * be downloaded on every test run. By specification, it cannot change
+	 * and will not be updated.
+	 *
+	 * @ticket 60698
```
As this is a helper rather than a test, the ticket number isn't necessary

dmsnell replied:
removed in 0e7ab4ca52d97da25a20e0e493bc33308f56cf1d

aaronjorbin commented on tests/phpunit/tests/wp-token-map/wpTokenMap.php:
```diff
@@ -0,0 +1,400 @@
+			$this->assertSame(
+				$replacement,
+				$response,
+				'Returned the wrong replacement value'
```
Minor: Make the error a bit more descriptive
```suggestion
				'Returned the wrong replacement value for ' . $token
```

dmsnell replied:
included the token in the message in 0e7ab4ca52d97da25a20e0e493bc33308f56cf1d

aaronjorbin reviewed (commented):
This is looking good. Thanks! 

One suggestion that can be done in a follow up, I think automating pulling https://html.spec.whatwg.org/entities.json and checking to see if the autogenerated file has been changed would be beneficial. Can likely be done on a weekly schedule and only in trunk.

dmsnell commented:
> One suggestion that can be done in a follow up, I think automating pulling https://html.spec.whatwg.org/entities.json and checking to see if the autogenerated file has been changed would be beneficial. Can likely be done on a weekly schedule and only in trunk.

this is a good suggestion, @aaronjorbin - and thanks for the review! (I'll be addressing the other feedback inline where you left it).

my one main reluctance to automate the download is that the specification requires that the list never change, and automating it only adds a point of failure to the process, well, a point of failure plus some latency.

do you think that it's worth adding this additional complexity in order to change what is defined to not change? I believe that a wide cascade of failures would occur around the internet at large if we found new character references in HTML.

I'm happy to add this if there are strong feelings about it, though if that's the case, maybe we could consider it a follow-up task so as not to delay this functionality? a separate task where we can discuss the merits of the automation?

dmsnell committed:
PR Feedback

dmsnell commented:
> I think we can add a new token-map group. It also might be worth putting in the html-api group since this is where the bigger effort lies

@aaronjorbin following the html5lib test suite, I added the `wpTokenMap` test module to the `@group html-api-token-map` group. happy to change if this isn't idea; I was looking for more examples and didn't see. it can also be set as a `@package` and `@subpackage` but I was looking for `@subgroup` and didn't find any

aaronjorbin commented:
> @aaronjorbin following the html5lib test suite, I added the `wpTokenMap` test module to the `@group html-api-token-map` group. happy to change if this isn't idea; I was looking for more examples and didn't see. it can also be set as a `@package` and `@subpackage` but I was looking for `@subgroup` and didn't find any

@subgroup isn't an annotation that is a part of PHPUnit.  I think this group makes sense, thanks!

aaronjorbin commented:
> > One suggestion that can be done in a follow up, I think automating pulling https://html.spec.whatwg.org/entities.json and checking to see if the autogenerated file has been changed would be beneficial. Can likely be done on a weekly schedule and only in trunk.
> 
> this is a good suggestion, @aaronjorbin[<img alt="" width="12" height="12" src="chrome-extension://hgomfjikakokcbkjlfgodhklifiplmpg/images/wp-logo.png">](https://profiles.wordpress.org/jorbin) - and thanks for the review! (I'll be addressing the other feedback inline where you left it).
> 
> my one main reluctance to automate the download is that the specification requires that the list never change, and automating it only adds a point of failure to the process, well, a point of failure plus some latency.
> 
> do you think that it's worth adding this additional complexity in order to change what is defined to not change? I believe that a wide cascade of failures would occur around the internet at large if we found new character references in HTML.
> 
> I'm happy to add this if there are strong feelings about it, though if that's the case, maybe we could consider it a follow-up task so as not to delay this functionality? a separate task where we can discuss the merits of the automation?

Upon reading https://github.com/whatwg/html/blob/main/FAQ.md#html-should-add-more-named-character-references I withdrawal this suggestion. I do think it might make sense to link that and mention that the entities json should never change in the readme for others that come across this with a similar thinking as me.

dmsnell commented:
> I do think it might make sense to link that and mention that the entities json should never change in the readme for others that come across this with a similar thinking as me.

Thanks again @aaronjorbin!

It's already mentioned at the [top of the README](https://github.com/WordPress/wordpress-develop/pull/5373/files#diff-315b92243d520f7e335b19bc91cce445d30fe57b5983524d01d804f060cf4dd1R8), in [the character reference class](https://github.com/WordPress/wordpress-develop/pull/5373/files#diff-f8b11c8be458ccb1fc5b53d12c9a56886c0bd9e0dd9ed1551a78a1c716aaae93R26) (and as well in [the autogenerator script](https://github.com/WordPress/wordpress-develop/pull/5373/files#diff-c4525aae4ea3b04a7b228928d4bc2eee144e4482ef1e1fbfc3d0ebce1702271dR76-R86) where it comes from), the PR description, and the Trac ticket description.

Is that good enough or was there somewhere else you wanted it? Maybe the way I wrote it wasn't as direct or clear as it should be?

aaronjorbin reviewed (approved):
This is looking good to me. Thanks for answering all my questions @dmsnell!

dmsnell committed:
Merge branch 'trunk' into experiment/add-wp-token-set

dmsnell commented:
Merged in [[58188]](https://core.trac.wordpress.org/changeset/58188)
https://github.com/wordpress/wordpress-develop/commit/bcd25b14ec0208657af1299b6df5642b5c0260a9

sirreal commented on src/wp-includes/class-wp-token-map.php:
```diff
@@ -0,0 +1,818 @@
+
+		// Longer strings are less-than for comparison's sake.
+		if ( $length_a !== $length_b ) {
+			return $length_b - $length_a;
```
This is a perfect place for the spaceship operator comparison to return -1, 0, or 1 `return $length_b <=> $length_a;`

dmsnell replied:
are we sure about this?

we need to make sure that longer strings sort before shorter ones, so for example, `ba` appears before `a`

```php
-1 === longest_first_then_alphabetical( 'ba', 'a' );
 1 === 'ba' <=> 'a';
 1 === strcmp( 'ba', 'a' );
```

sirreal replied:
Yes, I think this is correct. I meant specifically the length comparison could be `return $length_b <=> $length_a;`. The result is basically the same, but it returns `-1`, `0`, or `1` instead of `<0`, `0`, or `>0`.

<details>

<summary>demo</summary>

```php
<?php

function longest_first_then_alphabetical( $a, $b ) {
	if ( $a === $b ) {
		return 0;
	}


	$length_a = strlen( $a );
	$length_b = strlen( $b );


	// Longer strings are less-than for comparison's sake.
	if ( $length_a !== $length_b ) {
		return $length_b - $length_a;
	}


	return strcmp( $a, $b );
}
	
function longest_first_then_alphabetical_spaceship( $a, $b ) {
	if ( $a === $b ) {
		return 0;
	}


	$length_a = strlen( $a );
	$length_b = strlen( $b );


	// Longer strings are less-than for comparison's sake.
	if ( $length_a !== $length_b ) {
		// 👇👇👇 This is the only different line 👇👇👇
		return $length_b <=> $length_a;
	}


	return strcmp( $a, $b );
}
	
	
var_dump(
    longest_first_then_alphabetical( 'baa', 'a'),
    longest_first_then_alphabetical_spaceship( 'baa', 'a'),
    "\n",

    longest_first_then_alphabetical( 'abb', 'a'),
    longest_first_then_alphabetical_spaceship( 'abb', 'a'),
    "\n",

    longest_first_then_alphabetical( 'a', 'abb'),
    longest_first_then_alphabetical_spaceship( 'a', 'abb'),
    "\n",

    longest_first_then_alphabetical( 'a', 'abb'),
    longest_first_then_alphabetical_spaceship( 'a', 'abb'),
    "\n",

    longest_first_then_alphabetical( 'a', 'baa'),
    longest_first_then_alphabetical_spaceship( 'a', 'baa'),
    "\n",

    longest_first_then_alphabetical( 'a', 'baa'),
    longest_first_then_alphabetical_spaceship( 'a', 'baa'),
    "\n",

    longest_first_then_alphabetical( 'a', 'a'),
    longest_first_then_alphabetical_spaceship( 'a', 'a'),
    "\n",

    longest_first_then_alphabetical( 'b', 'a'),
    longest_first_then_alphabetical_spaceship( 'b', 'a'),
    "\n",

    longest_first_then_alphabetical( 'a', 'b'),
    longest_first_then_alphabetical_spaceship( 'a', 'b'),
    "\n",

);
```
int(-2)
int(-1)
string(1) "
"
int(-2)
int(-1)
string(1) "
"
int(2)
int(1)
string(1) "
"
int(2)
int(1)
string(1) "
"
int(2)
int(1)
string(1) "
"
int(2)
int(1)
string(1) "
"
int(0)
int(0)
string(1) "
"
int(1)
int(1)
string(1) "
"
int(-1)
int(-1)
string(1) "
"
```

</details>
